tests/std/tests/P0408R7_efficient_access_to_stringbuf_buffer/test.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <memory_resource>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include <test_death.hpp>
+
+using namespace std;
+
+constexpr const char empty_string[] = "";
+constexpr const char small_string[] = "a";
+constexpr const char large_string[] =
+    "This is a long long long long long long long long string to avoid small string optimization.";
+
+template <typename Stream>
+struct test_rvalue {
+    void operator()(const string& init_value) {
+        string buffer{init_value};
+        Stream stream{move(buffer)};
+        assert(stream.view() == init_value);
+        assert(stream.str() == init_value);
+        assert(stream.view() == init_value);
+        assert(stream.rdbuf()->get_allocator() == init_value.get_allocator());
+        // Move out the buffer, the underlying buffer should be empty.
+        buffer = move(stream).str();
+        assert(buffer == init_value);
+        assert(stream.view().empty());
+        assert(stream.str().empty());
+        // Move in the buffer string
+        stream.str(move(buffer));
+        assert(stream.view() == init_value);
+        assert(stream.str() == init_value);
+        assert(stream.rdbuf()->get_allocator() == init_value.get_allocator());
+        // Move to another stream
+        Stream stream2 = move(stream);
+        assert(stream.view().empty());
+        assert(stream2.view() == init_value);
+    }
+};
+
+template <typename Stream>
+struct test_allocator {
+    void operator()(const pmr::string& init_value) {
+        Stream stream{init_value};
+        assert(stream.view() == init_value);
+        assert(stream.str(init_value.get_allocator()) == init_value);
+        assert(stream.str() == string_view{init_value});
+        assert(move(stream).str() == string_view{init_value});
+        assert(stream.view().empty());
+        stream.str(init_value);
+        assert(stream.view() == init_value);
+        assert(stream.str(init_value.get_allocator()) == init_value);
+        assert(stream.str() == string_view{init_value});
+    }
+};
+
+template <typename Test>
+void run_test_util() {
+    Test test{};
+    test(empty_string);
+    test(small_string);
+    test(large_string);
+}
+
+template <template <typename> typename Test>
+void run_test() {
+    run_test_util<Test<stringstream>>();
+    run_test_util<Test<istringstream>>();
+    run_test_util<Test<ostringstream>>();
+}
+
+template <typename T>
+struct counting_allocator {
+    using value_type = T;
+
+    std::allocator<T> allocator{};
+    std::size_t count{0};

```

Comment:
It seems like this should be `shared_ptr<size_t>` so that all the allocator copies share the same counter?

Reply:
I tried, but it caused some problems, maybe because of using of rebind allocators. The current implementation seems OK.

Reply:
But the current implementation does not catch the bug you want to catch because the counter is always going to be 1 since the original allocator is left untouched.